babel/number_spelling.py

```diff
@@ -0,0 +1,549 @@
+        """
+        Search for digit function in the module based on data
+        """
+        self._speller = self.__class__.get_spell_function(locale)
+
+    def apply(self, number, ordinal=False, **kwargs):
+        """
+        Apply the speller
+        """
+        # generate the number if a ContextMaker exists
+        if hasattr(self._speller, '_context_maker'):
+            number = self._speller._context_maker(number)
+
+        # call transform
+        self._speller(number, ordinal)
+
+        # return the result (construct it recursively)
+        return '{0}'.format(number)
+
+    @classmethod
+    def get_spell_function(cls, locale):
+        """
+        Search `cls.registry` for the spelling function
+
+        NEgotiation is called in `babel.numbers.spell_number`
+        """
+        if locale in cls.registry:
+            return cls.registry[locale]
+        else:
+            raise SpellerNotFound("There is no appropriate number speller function for the given locale")
+
+
+    @classmethod
+    def decorate_spell_function(cls, locale, context_maker=None):
+        """
+        This is the parameterized decorator for the spellers
+        """
+        def decorator(func):
+            """
+            Add the speller function to registry
+            """
+            # add contextmaker to function if not empty
+            if context_maker is not None:
+                func._context_maker = context_maker
+
+            # check for override issue warning
+            cls.registry[locale] = func
+            # pervent the module functions to be called directly -- return `None`
+            return None
+
+        return decorator
+
+
+# shortcut for cleaner code in the locale specific spell function definitions
+spell = NumberSpeller.decorate_spell_function
+
+
+class ContextMaker(object):
+    """
+    Various complecated grouping mechanism could be implemented
+    with different `ContextMaker` classes.
+
+    If no `ContextMaker` is provided for the decorator of a speller
+    function then the speller is called with the raw number.
+    """
+    def __init__(self, integer_grouping=3, fraction_grouping=3, precision=6):
+        """
+        The defaults are based on the Hungarian speller but specified
+        there explicitly.
+        """
+        self.integer_grouping = integer_grouping
+        self.fraction_grouping = fraction_grouping
+        self.precision = precision
+
+    def __call__(self, number):
+        """
+        Process number and return context
+
+        This is a basic implementation of grouping.
+        """
+        def reverse_group_by(digits, group_size):
+            digits.reverse()
+            groups = []
+            while len(digits) > 0:
+                groups.append(digits[:group_size])
+                digits = digits[group_size:]
+            return groups
+
+        decimal.getcontext().prec = 99 # could be more or less but it should be as many as the supported digits
```

Comment:
This approach is not optimal since it changes the global (per thread) decimal context, which can break user code that relies on decimal context stability, in case a user changes it.

The best way to solve this is to use a decimal context manager. For example:

https://github.com/etanol/babel/commit/7ced0273020427cad3bc321001b64f769cf2f85d